File: skellig-task/src/main/java/org/skellig/task/TaskUtils.java
package org.skellig.task;


import org.skellig.task.exception.TaskRunException;

import java.util.concurrent.Callable;
import java.util.function.Predicate;

public final class TaskUtils {

    private static final double MS_CONSTANT = 1000;

    private TaskUtils() {
    }

    public static <T> T runTask(Callable<T> task, Predicate<T> stopCondition, int delay, int timeout) throws TaskRunException {
        return runTaskUntil(task, stopCondition, delay, timeout, null);
    }

    private static <T> T runTaskUntil(Callable<T> task, Predicate<T> stopCondition, int delay, double timeout, T previousResult) throws TaskRunException {
        if (timeout >= 0) {
            long startTime = System.currentTimeMillis();
            T result;
            try {
                result = task.call();
            } catch (Exception ex) {
                throw new TaskRunException(ex.getMessage(), ex);
            }
            int totalExecutionTime = (int) (System.currentTimeMillis() - startTime);

            if (stopCondition.test(result)) {
                return result;
            } else {
                if (delay > 0) {
                    delay(delay);
                }
                return runTaskUntil(task, stopCondition, delay, timeout - delay/MS_CONSTANT - totalExecutionTime, result);
            }
        }
        return previousResult;
    }

    private static void delay(int idleMilliseconds) {
        try {
            Thread.sleep(idleMilliseconds);
        } catch (InterruptedException ignored) {
        }
    }
}

File: skellig-task/src/main/kotlin/org/skellig/task/TaskUtils.kt
package org.skellig.task

import java.util.concurrent.Callable
import java.util.function.Predicate

class TaskUtils {

    companion object {

        @JvmStatic
        @Throws(TaskRunException::class)
        fun <T> runTask(task: Callable<T>, stopCondition: Predicate<T>, delay: Int, timeout: Int): T {
            return runTaskUntil(task, stopCondition, delay, timeout)
        }

        @JvmStatic
        @Throws(TaskRunException::class)
        private fun <T> runTaskUntil(task: Callable<T>, stopCondition: Predicate<T>, delay: Int, timeout: Int): T {
            var newTimeout: Int
            var result: T
            do {
                val startTime = System.currentTimeMillis()
                result = try {
                    task.call()
                } catch (ex: Exception) {
                    throw TaskRunException(ex.message, ex)
                }
                val totalExecutionTime = (System.currentTimeMillis() - startTime).toInt()
                newTimeout = if (stopCondition.test(result)) {
                    0
                } else {
                    if (delay > 0) {
                        delay(delay)
                    }
                    timeout - delay - totalExecutionTime
                }
            } while (newTimeout > 0)

            return result
        }

        private fun delay(idleMilliseconds: Int) {
            try {
                Thread.sleep(idleMilliseconds.toLong())
            } catch (ignored: InterruptedException) {
            }
        }
    }
}